### ISSUE TITLE
Define push model for notebook kernels

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#119949]
Today, notebook kernels are operating on a pull model, e.g there is a kernel provider that returns kernels. With that approach there is an API violation since kernels aren't data-only objects and the overall the pull model causes headaches because we don't really know when to pull for kernels. Instead, kernels should use a push model, e.g extensions create kernels as managed objects and are responsible for their lifecycle.  

This means there will be a `createNotebookKernel`-function which returns a kernel instance. The kernel will have properties and functions that allow to drive the VS Code UI for a kernel (this is similar to `createStatusBarItem` or `createSourceControl`). This leads to the question what a kernel is? As far as VS Code is concerned kernels enable the execute commands, show in the status bar, and drive the kernel picker. There is no further mapping of "actual" kernels, the API is there to drive the UI and initiate execution. 

This leads to an API like this (rough sketches!)

```ts

export function createNotebookKernel(id: string, label: string, selector: NotebookSelector, executeHandler: (notebook: NotebookDocument, cells: NotebookCell[]) => void): NotebookKernel;

export interface NotebookKernel {

  readonly id: string;

  // select notebook of a type and/or by file-pattern
  readonly selector: NotebookSelector;

  // is this kernel selected
  readonly selected: boolean;

  // fired when kernel is selected/unselected
  readonly onDidChangeSelection: Event<boolean>;

  // UI properties (get/set)
  name: string;
  description: string;
  supportedLanguages: string[];  
  hasExecutionOrder: boolean;

  // invoked when Run, Run All, Run Selections is triggered,
  // command is invoked with [kernel, cells] as arguments
  executeHandler: (notebook: NotebookDocument, cells: NotebookCell[]) => void

  // optional kernel interrupt command
  interruptHandler?: (notebook: NotebookDocument) => void
 
  // kernels (and _only_ they) can create executions
  createNotebookCellExecutionTask(uri: Uri, index: number): NotebookCellExecutionTask;

  // kernels can establish IPC channels to (visible) notebook editors
  createNotebookCommunication(editor: NotebookEditor): NotebookCommunication;

  // remove kernel
  dispose(): void;
}

```

Details are to be defined but gist of it should be clear. Consequences that are not so obvious are the following: 

* when and iff kernels appear depends entirely on extensions and we are at the mercy of extensions getting this right, e.g be swift and painless in creating kernels (good defaults, no prompts etc). This is not novel and seems to work well enough for similar APIs, like SCM or comments.
* Not strictly a consequence of this but kernels should become "private" objects, e.g only the extension that creates a kernel knows about it and no public object can reference a kernel. That means `NotebookCommunication#kernel` and `onDidChangeActiveNotebookKernel` should go away. Already today that's fishy because any extension can interact with any kernel, like interrupting it or executing arbitrary cells.

**Comment by jrieken** [comment]
> This leads to the question what a kernel is? As far as VS Code is concerned kernels enable the execute commands, show in the status bar, and drive the kernel picker. There is no further mapping of "actual" kernels, the API is there to drive the UI and initiate execution.

This should allow/encourage us to think about a different name. Maybe just `NotebookController` (similar to SourceControl and CommentController)

**Comment by jrieken** [comment]
This should address: https://github.com/microsoft/vscode/issues/117077, https://github.com/microsoft/vscode/issues/117506, https://github.com/microsoft/vscode/issues/118280

**Comment by mjbvz** [comment]
I'm not super familiar with current state of the notebook api so apologies if these are basic questions:

1. When does an extension call `createKernel `? Would there be an activation event for kernels?

2. Would the extension now control the entire lifecycle of the kernel? Are there any cases where VS Code would need to dispose of a kernel?
1. If we show kernels in the UI, would there be a kernel contribution in the `package.json`?
1. Should properties like `executeCommand` instead be passed in to the `createKernel ` factory function? Or is there a use case for changing these properties after the kernel is created?
2. Instead of using commands for `executeCommand` and `interruptCommand `, could we pass implementations of these functions to `createKernel`?

**Comment by jrieken** [comment]
> When does an extension call createKernel? Would there be an activation event for kernels?

We have an event for notebooks, `onNotebook:<type>` and my expectation is that extensions adding kernels use the same. 

> Would the extension now control the entire lifecycle of the kernel? Are there any cases where VS Code would need to dispose of a kernel?

I don't believe so - for us a kernel is an entry in the kernel picker and the thing behind the execute button. We don't really want more from it. Some kernels are simple (like GH issue notebooks) and some are very complex, like jupyter notebooks. Since we cannot properly generalise anything here we make the API a simple UI-driver. Currently, we only show kernels when having a notebook editor active and for us kernels aren't expensive, e.g the don't allocate a real resource like webviews


> Should properties like executeCommand instead be passed in to the createKernel factory function? Or is there a use case for changing these properties after the kernel is created?

Yeah, maybe. Tho, I believe having mutable commands isn't hurting us so I would not prevent them. 

> Instead of using commands for executeCommand and interruptCommand , could we pass implementations of these functions to createKernel?

The `SourceControl` type works with commands and so do others, like status bar items and code lenses. The [`CommentController.reactionHandler`](https://github.com/microsoft/vscode/blob/996391b87d3b71393d841bd0d3729a147418c7b2/src/vs/vscode.d.ts#L12410-L12410) took a different, more typed, approach. The command lacks typing but I like that it allows for some UI control, e.g the execute-command can show a different label per kernel or kernel state

**Comment by DonJayamanne** [comment]
**Some concerns/issues with push model**
* How does VS Code pick a kernel if extension does't provide a preferred kernel & extension has only provided 5 kernels & still busy fetching more...
* How will VS Code sort the list of the kernels (will there be a property or the like on Kernel to indicate the fact that its used for sorting)

**Basically the requirement is:**
* Extension should be able to provide a preferred kernel ASAP, so VSCode displays this immediately
	* This way user can run cell, even if all other 100 kernels have not been discovered.
* If user tries to select a kernel, let them know that extension is still busy searching/fetching kernels

**Comment by jrieken** [comment]
> If user tries to select a kernel, let them know that extension is still busy searching/fetching kernels

Today we don't have good UI for that; we show progress in the kernel picker but pickers are easy to dismiss and therefore not a state in which I am willing to wait for a longer time. The question is wether a dedicated UI (and with that API) is needed or not. E.g jupyter could use the `window.withProgress`-API to tell users what it is doing and no further API would be needed.

**Comment by jrieken** [comment]
further feedback from discussion: 

* allow kernels to be created with a `NotebookDocumentFilter` so that they can target a specific notebook document or all notebooks of a type
* add a property that indicates if a kernel sets the execution order (replacement for [hasExecutionOrder](https://github.com/microsoft/vscode/blob/29d59497cd3eecf136903ed971ef32e75913766f/src/vs/vscode.proposed.d.ts#L979-L979))

**Comment by firelizzard18** [comment]
IMO there should be an option for notebooks to have multiple active kernels. For example, the [markdown notebook](https://github.com/microsoft/vscode-markdown-notebook) extension defines a notebook format (content provider) but does not define any kernels, and my [Go kernel](https://gitlab.com/ethan.reesor/vscode-notebooks/go-kernel) and [HTTP kernel](https://gitlab.com/ethan.reesor/vscode-notebooks/http-kernel) each define a notebook kernel but no format.

In cases like the markdown notebook format and domain-specific kernels, only having a single active kernel is rather limiting.

**Comment by jrieken** [comment]
> IMO there should be an option for notebooks to have multiple active kernels. 

You can have multiple kernels but I don't know how multiple active kernels should work? For us, the active kernel is the one that executes cells. So, maybe you can have a kernel per cell but how should multiple kernels for one cell work? How do I pick which one runs? All? Make run a dropdown? Does the output of each kernel merge?

**Comment by firelizzard18** [comment]
> So, maybe you can have a kernel per cell

That's what I meant - the possibility of setting the kernel per cell. I'm not suggesting multiple kernels for a given cell. It may not make sense for Jupyter notebooks, so maybe "allow a kernel per cell" should be specified by the content provider. But for generic notebooks like markdown, I'd like the ability to execute Go blocks against the Go kernel, and other blocks against other kernels.

I wonder if it would be worthwhile to have some kind of indication that "this content/kernel provider is generic". For example, my Go kernel will happily execute Go code regardless of what kind of notebook it is in - it has no dependencies on the actual notebook content/format/provider. I tried registering it with `{ viewType: ['go'] }` (with no filename pattern), but that didn't work. I thought about registering it as `{ viewType: ['go'], filenamePattern: '*' }`, but that didn't seem like the best idea. It might be useful to be able to register a kernel as "this kernel will run code from any notebook (as long as the notebook doesn't have specific expectations of the kernel)". And vise versa, "this content provider will work with any kernel (sans specific expectations)". That would allow differentiation between tightly coupled notebooks and generic content/kernel providers.

Long term, I'd like to develop an extension for querying databases. If the active kernel could be set per cell, I could register a kernel per database connection, which would allow the user to set the database connection for a SQL cell by setting the kernel for that cell. That way I could avoid things like the comment directives that .NET interactive notebooks use to set the connection.

**Comment by DonJayamanne** [comment]
>   // UI properties (get/set)
  name: string;
  description: string;

This is awesome, I'm planning on using this in a separate extension. 
Basically have the same kernel, but change the display name (display cluster & database) during run time.  https://github.com/donjayamanne/vscode-kusto

Here's how I plan on hacking the UI:
* Add a single item <Configure Kusto Connection>
* When the user selects this from the kernel quick pick, I'll:
	* Hook into the kernel selection changed event & see that the user chose this option
	* Now i'll display my own quick pick asking the user to select a cluster & db (using https://github.com/microsoft/vscode-extension-samples/blob/main/quickinput-sample/src/multiStepInput.ts#L7)
* Once user selects a cluster & database, I update the display name & description of the kernel to the real values & then I'll contribute a new kernel that'll show up as `<Configure Kusto Connection>`
* This way user can still use the kernel picker to add/configure new connections.

Basically its a hack to introduce my own kernel picker (hacky).

Thought I'd let u know how about my work around to introducing my own kernel picker, hint -> Asking for an event that gets triggered for **Extensions to handle their own kernel pickers.**

This would work just great Jupyter Extension as well (we'd like to have our own kernel picker)
See here for some previous work that @misolori did in this space https://github.com/microsoft/vscode/issues/106519

**Comment by DonJayamanne** [comment]
> Asking for an event that gets triggered for Extensions to handle (display) their own kernel pickers.

Not sure this will work, when other extension authors contribute kernels for notebooks.

**Comment by jrieken** [comment]
> I wonder if it would be worthwhile to have some kind of indication that "this content/kernel provider is generic".

Our plans is that all content providers are generic - we are working on decoupling it from all execution related information. 

> For example, my Go kernel will happily execute Go code regardless of what kind of notebook it is in - it has no dependencies on the actual notebook content/format/provider. 

Yeah, that's our idea as well. Kernels are decoupled from content providers but for us notebook cells don't live in isolation from each other and are executed by the same kernel. Tho, you can achieve what you want by doing the indirection yourself: have one (vscode) kernel that support all/many languages and that in return dispatches to different "executors". 

> I tried registering it with { viewType: ['go'] } (with no filename pattern), but that didn't work. I thought about registering it as { viewType: ['go'], filenamePattern: '*' }

The `viewType` property is the notebook type, not the language identifier of a code cell. To target certain code cells you need to use the `supportedLanguages` property

**Comment by jrieken** [comment]
fyi - a first version of this exists (and works) inside this branch `joh/kernel-push`

I have also played with some toy kernels and one change is that execution and interrupt isn't done via command anymore but via callback. I have updated the proposal above. One thought on this is that we could consider to pass cell execution objects to the execute handler (since we expect them to be created anyways)

**Comment by DonJayamanne** [comment]
```typescript

		// ipc
		readonly preloads: NotebookKernelPreload[];
		readonly onDidReceiveMessage: Event<{ editor: NotebookEditor, message: any }>;
		postMessage(message: any, editor?: NotebookEditor): Thenable<boolean>;
		asWebviewUri(localResource: Uri): Uri;
```

The questions are around Notebook Controllers:
* `postMessage(message: any, editor?: NotebookEditor): Thenable<boolean>;`
If `editor` is undefined, do we send the message to all editors associated with the same document or all editors (even from other dociuments).
We would need a way to send messages to all editors related to a single document.
**This is required for IPyWidgets**

* `asWebviewUri(localResource: Uri): Uri;`
Does this work with split editors?
Assume i open a.ipynb, and thens split it.
would the webviews used to display output in both editors (pointing to the same notebook document) end up with the exact same scheme? 
I'm assuming yes, as ther'es no additional argument here.

**Comment by jrieken** [comment]
> If editor is undefined, does VS Code send this message to all editors associated with the same document related to this kernel or woudl it get sent to all Notebook editor in VS Code (even if its not associated with this kernel).

Yeah, it will be all editors unless one is "picked". Also yeah, postMessage will be ignored (it will resolve to `false`) when a kernel isn't "bound" to the notebook the editor is showing.

> I'm assuming yes, as there's no additional argument here.

I am assuming the same 😆 The editor was only used to seed an identifier when generating the URI. The actual form isn't honoured anywhere (afaik) and therefore I decided its easier if we do that - we already have a unique handle per controller that we are now using.

**Comment by jrieken** [comment]
Closing this as we shipped the new notebook controller API. Feedback welcome, but preferable in separate topic issues
